src/main/windows/SDL_sysmain_runapp.c

```diff
@@ -20,80 +20,104 @@
-  based on SDL_windows_main.c, placed in the public domain by Sam Lantinga  4/13/98 */
-
 #include <shellapi.h> // CommandLineToArgvW()
 
-// Pop up an out of memory message, returns to Windows
 static int OutOfMemory(void)
 {
     SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_ERROR, "Fatal Error", "Out of memory - aborting", NULL);
     return -1;
 }
 
-int MINGW32_FORCEALIGN SDL_RunApp(int _argc, char* _argv[], SDL_main_func mainFunction, void * reserved)
+static int ErrorProcessingCommandLine(void)
 {
-    /* Gets the arguments with GetCommandLine, converts them to argc and argv
-       and calls SDL_main */
-
-    LPWSTR *argvw;
-    char **argv;
-    int i, argc, result;
-
-    (void)_argc; (void)_argv; (void)reserved;
-
-    argvw = CommandLineToArgvW(GetCommandLineW(), &argc);
-    if (!argvw) {
-        return OutOfMemory();
-    }
-
-    /* Note that we need to be careful about how we allocate/free memory here.
-     * If the application calls SDL_SetMemoryFunctions(), we can't rely on
-     * SDL_free() to use the same allocator after SDL_main() returns.
-     */
+    SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_ERROR, "Fatal Error", "Error processing command line arguments", NULL);
+    return -1;
+}
 
-    // Parse it into argv and argc
-    argv = (char **)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, (argc + 1) * sizeof(*argv));
-    if (!argv) {
-        return OutOfMemory();
-    }
-    for (i = 0; i < argc; ++i) {
-        const int utf8size = WideCharToMultiByte(CP_UTF8, 0, argvw[i], -1, NULL, 0, NULL, NULL);
-        if (!utf8size) {  // uhoh?
-            SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_ERROR, "Fatal Error", "Error processing command line arguments", NULL);
-            return -1;
+// If the provided argv is NULL (which it will be when using SDL_main), this function parses
+// the command-line string for the current process into an argv and uses that instead.
+// Otherwise, the provided argv is used as-is (since that's probably what the user wants).
+int SDL_CallMain(int caller_argc, char* caller_argv[], SDL_main_func mainFunction)
+{
+    int result, argc;
+    LPWSTR *argvw = NULL;
+    char **argv = NULL;
+
+    // Note that we need to be careful about how we allocate/free memory in this function. If the application calls
+    // SDL_SetMemoryFunctions(), we can't rely on SDL_free() to use the same allocator after SDL_main() returns.
+
+    if (!caller_argv || caller_argc < 0) {
+        // If the passed argv is NULL (or argc is negative), SDL should get the command-line arguments
+        // using GetCommandLineW() and convert them to argc and argv before calling mainFunction().
+
+        // Because of how the Windows command-line works, we know for sure that the buffer size required to store all
+        // argument strings converted to UTF-8 (with null terminators) is guaranteed to be less than or equal to the
+        // size of the original command-line string converted to UTF-8.
```

Comment:
See https://learn.microsoft.com/en-us/cpp/cpp/main-function-command-line-args?view=msvc-170#parsing-c-command-line-arguments for details.

The gist is that the arguments in the command-line string are always either `bare`, `"quoted"` or `"escape \\"quotes\\" using backslashes"`, which means that when parsed, each argument always results in a string of the same length or shorter than the argument in the command-line string, never longer. In addition, arguments are always separated by at least one space character (and the entire command-line string is null-terminated), so if we use the length of the whole command-line string as the buffer size there's still enough room for null terminators.